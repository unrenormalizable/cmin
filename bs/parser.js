import * as scanner from './scanner.js'

export const AST_NODE_PROGRAM = 0
export const AST_NODE_FUNCTION = 0

export const PARSE_ACCEPTED = 0
export const PARSE_NOT_YET_ACCEPTED = 1
export const PARSE_NOT_ACCEPTED = 2

function readToken(state, tokenId, tokenValue) {
  if (state.scanState.token.id === tokenId && (!tokenValue || state.scanState.token.value === tokenValue)) {
    const scanState = scanner.scan(state.text, state.scanState.state)
    state.scanState = scanState
    state.result = PARSE_NOT_YET_ACCEPTED
  } else {
    state.result = PARSE_NOT_ACCEPTED
  }

  return state
}

function parseFunction(state) {
  state = { ...state }

  state = readToken(state, scanner.TOK_KEYWORD, 'fn')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  const fnAst = { id: AST_NODE_FUNCTION, name: null, return: 'void', statements: [] }
  fnAst.name = state.scanState.token.value
  fnAst.token = state.scanState.token

  state = readToken(state, scanner.TOK_IDENTIFIER)
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state = readToken(state, scanner.TOK_SYMBOL, '(')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state = readToken(state, scanner.TOK_SYMBOL, ')')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state = readToken(state, scanner.TOK_SYMBOL, '{')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state = readToken(state, scanner.TOK_SYMBOL, '}')
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state.ast.functions.push(fnAst)
  state.result = PARSE_NOT_YET_ACCEPTED
  return state
}

function parseFunctionList(state) {
  state = { ...state }

  if (state.scanState.token.id === scanner.TOK_EOF) {
    state.result = PARSE_NOT_YET_ACCEPTED
    return state
  }

  state = parseFunction(state)
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  return parseFunctionList(state)
}

function parseProgram(state) {
  state = { ...state }
  state.ast = { id: AST_NODE_PROGRAM, functions: [] }

  state = parseFunctionList(state)
  if (state.result === PARSE_NOT_ACCEPTED) {
    return state
  }

  state.result = PARSE_ACCEPTED
  return state
}

export function parse(text) {
  const scanState = scanner.scan(text, { pos: 0, line: 1, column: 1 })
  const state = { text, result: PARSE_NOT_YET_ACCEPTED, scanState }

  return parseProgram(state)
}
