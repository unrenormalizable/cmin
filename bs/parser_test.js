import * as assert from '@std/assert'
import { assertSnapshot } from '@std/testing/snapshot'
import * as parser from './parser.js'

Deno.test({
  name: 'parse: empty file',
  fn: async (tc) => {
    const text = ''
    const ret = parser.parse(text)
    assert.assertEquals(ret.result, parser.PARSE_ACCEPTED)
    await assertSnapshot(tc, ret, {})
  },
})

Deno.test({
  name: 'parse: empty file with whitespace',
  fn: async (tc) => {
    const text = '   \r\t  \r\n  \n  '
    const ret = parser.parse(text)
    assert.assertEquals(ret.result, parser.PARSE_ACCEPTED)
    await assertSnapshot(tc, ret, {})
  },
})

Deno.test({
  name: 'parse: random scanner error',
  fn: async (tc) => {
    const text = '   \r#  \r\n  \n  '
    const ret = parser.parse(text)
    assert.assertEquals(ret.result, parser.PARSE_NOT_ACCEPTED)
    await assertSnapshot(tc, ret, {})
  },
})

Deno.test({
  name: 'parse: trivial main function',
  fn: async (tc) => {
    const text = 'fn main() {\n}'
    const ret = parser.parse(text)
    assert.assertEquals(ret.result, parser.PARSE_ACCEPTED)
    await assertSnapshot(tc, ret, {})
  },
})

Deno.test({
  name: 'parse: multiple functions',
  fn: async (tc) => {
    const text = 'fn main() {\n}fn __puts() {}'
    const ret = parser.parse(text)
    assert.assertEquals(ret.result, parser.PARSE_ACCEPTED)
    await assertSnapshot(tc, ret, {})
  },
})
