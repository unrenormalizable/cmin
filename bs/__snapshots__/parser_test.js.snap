export const snapshot = {};

snapshot[`parse: empty file 1`] = `
{
  ast: {
    functions: [],
    id: 0,
  },
  result: 0,
  scanState: {
    state: {
      column: 1,
      line: 1,
      pos: 0,
    },
    token: {
      id: 1,
      location: {
        column: 1,
        line: 1,
        pos: 0,
      },
      value: null,
    },
  },
  text: "",
}
`;

snapshot[`parse: empty file with whitespace 1`] = `
{
  ast: {
    functions: [],
    id: 0,
  },
  result: 0,
  scanState: {
    state: {
      column: 3,
      line: 4,
      pos: 14,
    },
    token: {
      id: 1,
      location: {
        column: 3,
        line: 4,
        pos: 14,
      },
      value: null,
    },
  },
  text: "   \\r	  \\r
  
  ",
}
`;

snapshot[`parse: random scanner error 1`] = `
{
  ast: {
    functions: [],
    id: 0,
  },
  result: 2,
  scanState: {
    state: {
      column: 1,
      line: 2,
      pos: 4,
    },
    token: {
      id: 0,
      location: {
        column: 1,
        line: 2,
        pos: 4,
      },
      value: null,
    },
  },
  text: "   \\r#  \\r
  
  ",
}
`;

snapshot[`parse: trivial main function 1`] = `
{
  ast: {
    functions: [
      {
        id: 0,
        name: "main",
        return: "void",
        statements: [],
        token: {
          id: 4,
          location: {
            column: 4,
            line: 1,
            pos: 3,
          },
          value: "main",
        },
      },
    ],
    id: 0,
  },
  result: 0,
  scanState: {
    state: {
      column: 2,
      line: 2,
      pos: 13,
    },
    token: {
      id: 1,
      location: {
        column: 2,
        line: 2,
        pos: 13,
      },
      value: null,
    },
  },
  text: "fn main() {
}",
}
`;

snapshot[`parse: multiple functions 1`] = `
{
  ast: {
    functions: [
      {
        id: 0,
        name: "main",
        return: "void",
        statements: [],
        token: {
          id: 4,
          location: {
            column: 4,
            line: 1,
            pos: 3,
          },
          value: "main",
        },
      },
      {
        id: 0,
        name: "__puts",
        return: "void",
        statements: [],
        token: {
          id: 4,
          location: {
            column: 5,
            line: 2,
            pos: 16,
          },
          value: "__puts",
        },
      },
    ],
    id: 0,
  },
  result: 0,
  scanState: {
    state: {
      column: 16,
      line: 2,
      pos: 27,
    },
    token: {
      id: 1,
      location: {
        column: 16,
        line: 2,
        pos: 27,
      },
      value: null,
    },
  },
  text: "fn main() {
}fn __puts() {}",
}
`;
